blob/drivertest/drivertest.go

```diff
@@ -2734,6 +2737,70 @@ func testAs(t *testing.T, newHarness HarnessMaker, st AsTest) {
 	}
 }
 
+//func testIfNotExist(t *testing.T, newHarness HarnessMaker) {
+//	t.Helper()
+//
+//	const key = "blob-for-if-not-exist"
+//	const contents = "up and down"
+//
+//	ctx := context.Background()
+//	h, err := newHarness(ctx, t)
+//	if err != nil {
+//		t.Fatal(err)
+//	}
+//	defer h.Close()
+//	drv, err := h.MakeDriver(ctx)
+//	if err != nil {
+//		t.Fatal(err)
+//	}
+//	b := blob.NewBucket(drv)
+//	defer func() { _ = b.Close() }()
+//
+//	opts := blob.WriterOptions{
+//		ContentType: "text",
+//		IfNotExist:  true,
+//	}
+//
+//	// Create one file for the key
+//	w1, err := b.NewWriter(ctx, key, &opts)
+//	if err != nil {
+//		t.Fatal(err)
+//	}
+//
+//	defer func() {
+//		_ = b.Delete(ctx, key)
+//	}()
+//
+//	// Write to the file
+//	if _, err := w1.Write([]byte(contents)); err != nil {
+//		t.Fatal(err)
+//	}
+//
+//	// Closing the file (ie: thus writing the content to the file)
+//	// should not return an error
+//	if err := w1.Close(); err != nil {
+//		t.Fatal(err)
+//	}
+//
+//	// Create a new writer for the same key
+//	w2, err := b.NewWriter(ctx, key, &opts)
+//	if err != nil {
+//		t.Fatal(err)
+//	}
+//
+//	// Write to the file
+//	if _, err := w2.Write([]byte(contents)); err != nil {
+//		t.Fatal(err)
```

Comment:
The test needs to handle the case where Write returns the error, and also verify the PreconditionFailed.

Something like:

var err error
if _, err = w2.Write([]byte(contents)); err == nil {
  err = w2.Close()
}
if err == nil {
  t.Error("expected error rewriting key with IfNotExist, got nil")
}
if code := gcerrors.Code(err); code != gcerrors.PreconditionFailed {
t.Errorf("expected PreconditionFailed error, got %v", code)
}

(I believe this test fails right now for memblob, because you didn't return PreconditionFailed).

Reply:
Done! Good catch!